core/schema/cache.go

```diff
@@ -32,13 +34,68 @@ func (s *cacheSchema) Dependencies() []SchemaResolvers {
 }
 
 type cacheArgs struct {
-	Key string
+	Key       string
+	Namespace string `default:""`
 }
 
-func (s *cacheSchema) cacheVolume(ctx context.Context, parent *core.Query, args cacheArgs) (*core.CacheVolume, error) {
-	// TODO(vito): inject some sort of scope/session/project/user derived value
-	// here instead of a static value
-	//
-	// we have to inject something so we can tell it's a valid ID
-	return core.NewCache(args.Key), nil
+func (s *cacheSchema) cacheVolume(ctx context.Context, parent dagql.Instance[*core.Query], args cacheArgs) (dagql.Instance[*core.CacheVolume], error) {
+	var inst dagql.Instance[*core.CacheVolume]
+
+	if args.Namespace != "" {
+		return dagql.NewInstanceForCurrentID(ctx, s.srv, parent, core.NewCache(args.Key+args.Namespace))
+	}
+
+	m, err := parent.Self.Server.CurrentModule(ctx)
+	if err != nil && !errors.Is(err, core.ErrNoCurrentModule) {
+		return inst, err
+	}
+
+	namespaceKey := ""
+	if m != nil {
+		name, err := m.Source.Self.ModuleName(ctx)
+		if err != nil {
+			return inst, err
+		}
+
+		bk, err := m.Source.Self.Query.Buildkit(ctx)
+		if err != nil {
+			return inst, err
+		}
+
+		refString, err := m.Source.Self.RefString()
+		if err != nil {
+			return inst, err
+		}
+		parsedRef := parseRefString(ctx, bk, refString)
+
+		// use combination of name, modPath and subpath as namespace key.
+		// if we only use name there is a high chance for namespace key conflict
+		// and high chance of cache-miss if we use module source digest for namespace key
+		namespaceKey = name + parsedRef.modPath + parsedRef.repoRootSubdir
+	}
+
+	// if no namespace key, just return the NewCache based on key
+	if namespaceKey == "" {
+		return dagql.NewInstanceForCurrentID(ctx, s.srv, parent, core.NewCache(args.Key))
```

Comment:
nit: for the args to `core.NewCache` we should always do something like `namespaceKey + ":" + args.Key`.

We don't want there to be a possibility for a non-module call to get the same cache-key as a module api call - which currently it seems like it can.

Reply:
For a module call, this should never happen and should always have a namespaceKey populated. we are returning with just Key only if namespaceKey turns out to be empty.

Reply:
Yes, this is my point - if a non-module call (that's not using modules at all, by using the standalone SDK or even a standalone `dagger query`), it *still* shouldn't be able to access the same cache volumes as modules. They should be entirely separate.

Reply:
if we attach a random key as namespaceKey here for "non-module" scenario, I think that will suffice here?